### PR TITLE
Support SDK install when TMP_DIR is a on a different filesystem

### LINKED ISSUE #746: daml assistant installation fails if TMP_DIR is on a different file system
I just tried to install the new assistant on my personal machine and it fails with
```
Installing SDK release from directory.
daml: AssistantError {errContext = Just "Installing extracted SDK release.", errMessage = Just "Failed to move extracted SDK release to final location.", errInternal = Just "/tmp/daml-update-78b77cee27e84c2e/release: renameDirectory:renamePath:rename: unsupported operation (Invalid cross-device link)"}
```

The reason for this failure is that `/tmp` is a tmpfs filesystem on my machine and `renameDirectory` uses a `rename` syscall under the hood which fails if source and target are on different filesystems. It is possible to work around this by setting `TMP_DIR` to a directory in the same filesystem as `~/.daml`. However, having `/tmp` be a tmpfs is sufficiently common on Linux, that this is something that we should probably address in the assistant by falling back to a non-atomic copy + delete.

cc @associahedron

### PR DESCRIPTION
**Pull request by cocreature**
fixes #746

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-assistant/daml-project-config/DAML/Project/Util.hs [hunk 1]
```diff
@@ -3,10 +3,25 @@
 
 module DAML.Project.Util where
```
Comment (neil-da) [inline daml-assistant/daml-project-config/DAML/Project/Util.hs]: I love export lists...

File: daml-assistant/src/DAML/Assistant/Install.hs [hunk 2]
```diff
@@ -138,13 +141,24 @@ installExtracted env@InstallEnv{..} sourcePath =
...
             requiredIO "Failed to move extracted SDK release to final location." $
-                renameDirectory (unwrapSdkPath sourcePath) (unwrapSdkPath targetPath)
+                moveDirectory (unwrapSdkPath sourcePath) (unwrapSdkPath targetPath)
 
         requiredIO "Failed to set file mode of installed SDK directory." $
             setSdkFileMode (unwrapSdkPath targetPath)
 
         whenActivate env $ activateDaml env targetPath
 
+moveDirectory :: FilePath -> FilePath -> IO ()
```
Comment (neil-da) [inline daml-assistant/src/DAML/Assistant/Install.hs]: This probably belongs at the same place as copyDirectory.

File: daml-assistant/daml-project-config/DAML/Project/Util.hs (hunk 1 above)
Comment (cocreature) [reply daml-assistant/daml-project-config/DAML/Project/Util.hs]: Can we hlint for that?

Comment (neil-da) [reply daml-assistant/daml-project-config/DAML/Project/Util.hs]: Yes! If there's broad support I'll do it. You just add `- warn: {name: Use explicit module export list}` in the `.hlint.yaml`

Comment (associahedron) [reply daml-assistant/daml-project-config/DAML/Project/Util.hs]: I'm in favor!

Comment (cocreature) [reply daml-assistant/daml-project-config/DAML/Project/Util.hs]: :+1:

File: daml-assistant/daml-project-config/DAML/Project/Util.hs [hunk 3]
```diff
@@ -10,3 +22,26 @@ fromRightM f = either f pure
...
+copyDirectory :: FilePath -> FilePath -> IO ()
+copyDirectory src target = do
+    files <- listFilesRecursive src
+    forM_ files $ \file -> do
+        let baseName = makeRelative src file
+        let targetFile = target </> baseName
+        createDirectoryIfMissing True (takeDirectory targetFile)
+        copyFile file targetFile
+        p <- getPermissions targetFile
+        setPermissions targetFile p { writable = True }
```
[the remark this answers is not in the document]
Comment (cocreature) [reply daml-assistant/daml-project-config/DAML/Project/Util.hs]: Good catch, I had forgotten about this change. I’ve removed it from `copyDirectory` and moved it to `daml new`.

**Pull request merged**